src/v/archival/archiver_manager.cc

```diff
@@ -0,0 +1,1042 @@
+      ss::lw_shared_ptr<ntp_archiver> archiver) noexcept {
+        try {
+            vlog(_ctxlog.debug, "stopping and disposing ntp_archiver");
+            if (archiver == nullptr) {
+                return ss::now();
+            }
+            archiver->notify_leadership(model::node_id());
+            _upload_housekeeping.deregister_jobs(
+              archiver->get_housekeeping_jobs());
+            return archiver->stop()
+              .then_wrapped([this](ss::future<> fut) {
+                  if (fut.failed()) {
+                      auto err = fut.get_exception();
+                      vlog(
+                        _ctxlog.error, "ntp_archiver shutdown failure {}", err);
+                      return ss::make_exception_future<>(err);
+                  }
+                  vlog(_ctxlog.debug, "ntp_archiver stopped");
+                  return fut;
+              })
+              .finally([a = archiver] {});
+        } catch (...) {
+            vlog(
+              _ctxlog.error,
+              "ntp_archiver unexpected shutdown exception {}",
+              std::current_exception());
+            return ss::make_exception_future<>(std::current_exception());
+        }
+    }
+
+    model::ntp _ntp;
+    model::node_id _self_id;
+    ss::lw_shared_ptr<cluster::partition> _part;
+    ss::lw_shared_ptr<const archival::configuration> _config;
+    cloud_storage::remote& _remote;
+    cloud_storage::cache& _cache;
+    archival::upload_housekeeping_service& _upload_housekeeping;
+    ss::abort_source _as;
+    retry_chain_node _rtc;
+    retry_chain_logger _ctxlog;
+
+    // archiver (only if is_leader)
+    // note: this will be changed to ntp_archiver_api in the future
+    ss::lw_shared_ptr<ntp_archiver> _archiver;
+    // currently active background operation (start/stop archiver)
+    std::optional<ss::future<>> _bg_operation;
+    // barrier is used by stop method
+    std::optional<ss::promise<>> _barrier;
+};
+
+struct managed_partition : public managed_partition_fsm::state_machine_t {
+    managed_partition(
+      model::ntp ntp,
+      model::node_id broker_id,
+      ss::lw_shared_ptr<cluster::partition> part,
+      ss::lw_shared_ptr<const archival::configuration> config,
+      cloud_storage::remote& remote,
+      cloud_storage::cache& cache,
+      archival::upload_housekeeping_service& housekeeping)
+      : managed_partition_fsm::state_machine_t(
+        ntp,
+        broker_id,
+        std::move(part),
+        std::move(config),
+        remote,
+        cache,
+        housekeeping)
+      , _ntp(ntp)
+      , _node_id(broker_id) {}
+
+    ~managed_partition() {
+        auto st = current_state();
+        vassert(
+          st == managed_partition_state_t::passive,
+          "Unexpected managed_partition state {}",
+          st);
+    }
+
+    /// Acquired leadership notification
+    void notify_leadership_acquired(model::term_id term) {
+        process_event(ev_leadership_acquired{.term = term});
```

Comment:
In general I think it would have been very instructive if there were a test that exercised these `process_event()` calls directly. Then as a reviewer it's easier to validate that this change is correct, which is much harder to do without understanding boost.msm. Perhaps something to consider still

Reply:
I have such tests in the working branch but this PR is too big as is. Will add them later.

Reply:
Awesome! Those'll be great.